### PR TITLE
[Chat state layer] channel watchers

### PR DESCRIPTION
**Pull request by laevandus**
### 🔗 Issue Links

Related to [#728](https://github.com/GetStream/ios-issues-tracking/issues/728)

### 🎯 Goal

**Merges on top of [PR:3057](https://github.com/GetStream/stream-chat-swift/pull/3057)**

Adds the functionality in `ChatChannelWatcherListController` to the async await supported `Chat` + `ChatState`.

### 📝 Summary

- Chat
  - loadWatchers(with:)
  - loadMoreWatchers(with:)
- ChatState
  - watchers (property)
  - hasLoadedAllWatchers (property)
No unit-tests is expected since the final interface is WIP

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

File: Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift [hunk 1]
```diff
@@ -91,7 +91,8 @@ public class ChatChannelWatcherListController: DataController, DelegateCallable,
             return
         }
 
-        updater.channelWatchers(query: query) { error in
+        updater.channelWatchers(query: query) { result in
```
Comment (laevandus) [inline Sources/StreamChat/Controllers/ChannelWatcherListController/ChatChannelWatcherListController.swift:94]: I needed to access the returned channel payload

File: Sources/StreamChat/StateLayer/Chat.swift [hunk 2]
```diff
@@ -880,6 +887,28 @@ public final class Chat {
     public func stopWatching() async throws {
         try await channelUpdater.stopWatching(cid: cid)
     }
+    
+    /// Loads watchers for the specified pagination parameters and updates ``ChatState.watchers``.
```
Comment (martinmitrevski) [inline Sources/StreamChat/StateLayer/Chat.swift:891]: What's the difference between these 2 methods? The "more" part should be handled by the pagination, right?

File: Sources/StreamChat/Repositories/UserRepository.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        self.apiClient = apiClient
+    }
+}
+
+@available(iOS 13.0, *)
+extension UserRepository {
+    /// Returns an array of users in the local database.
+    func watchers(for ids: Set<UserId>, in cid: ChannelId) async throws -> [ChatUser] {
+        guard !ids.isEmpty else { return [] }
+        return try await database.backgroundRead { context in
```
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/UserRepository.swift:22]: Not in this scope, but food for thought: we should not save realtime or ephemeral stuff in the db, such as watchers, typing indicators, search results.

**Comment by laevandus** [comment]
Reviewed comments but closing since there are going to be larger changes around it.

**Pull request closed without merging**